jaxopt/_src/polyak_sgd.py

```diff
@@ -0,0 +1,187 @@
+    maxiter: maximum number of solver iterations.
+    tol: tolerance to use.
+    verbose: whether to print error on every iteration or not. verbose=True will
+      automatically disable jit.
+    implicit_diff: if True, enable implicit differentiation using cg,
+      if Callable, do implicit differentiation using callable as linear solver,
+      if False, use autodiff through the solver implementation (note:
+        this will unroll syntactic loops).
+    has_aux: whether ``fun`` outputs one (False) or more values (True).
+      When True it will be assumed by default that ``fun(...)[0]``
+      is the objective value. The auxiliary outputs are stored in
+      ``state.aux``.
+
+  References:
+    Berrada, Leonard and Zisserman, Andrew and Kumar, M Pawan.
+    "Training neural networks for and by interpolation".
+    International Conference on Machine Learning, 2020.
+
+    Loizou, Nicolas and Vaswani, Sharan and Laradji, Issam Hadj and
+    Lacoste-Julien, Simon.
+    "Stochastic polyak step-size for sgd: An adaptive learning rate for fast
+    convergence".
+    International Conference on Artificial Intelligence and Statistics, 2021.
+  """
+  fun: Callable
+  max_step_size: float = 1.0
+  delta: float = 0.0
+  momentum: float = 0.0
+  pre_update: Optional[Callable] = None
+  maxiter: int = 500
+  tol: float = 1e-3
+  verbose: int = 0
+  implicit_diff: Union[bool, Callable] = False
+  has_aux: bool = False
+
+  def init(self,
+           init_params: Any,
+           *args,
+           **kwargs) -> base.OptStep:
+    """Initialize the ``(params, state)`` pair.
+
+    Args:
+      init_params: pytree containing the initial parameters.
+      *args: additional positional arguments to be passed to ``fun``.
+      **kwargs: additional keyword arguments to be passed to ``fun``.
+    Return type:
+      base.OptStep
+    Returns:
+      (params, state)
+    """
+    del args, kwargs  # Not used.
+
+    if self.momentum == 0:
+      velocity = None
+    else:
+      velocity = tree_zeros_like(init_params)
+
+    state = PolyakSGDState(iter_num=0, error=jnp.inf, value=jnp.inf,
+                           aux=None, velocity=velocity)
+    return base.OptStep(params=init_params, state=state)
+
+  def update(self,
+             params: Any,
+             state: PolyakSGDState,
+             *args,
+             **kwargs) -> base.OptStep:
+    """Performs one iteration of the optax solver.
+
+    Args:
+      params: pytree containing the parameters.
+      state: named tuple containing the solver state.
+      *args: additional positional arguments to be passed to ``fun``.
+      **kwargs: additional keyword arguments to be passed to ``fun``.
+    Return type:
+      base.OptStep
+    Returns:
+      (params, state)
+    """
+    if self.has_aux:
+      (value, aux), grad = self._value_and_grad_fun(params, *args, **kwargs)
+    else:
+      value, grad = self._value_and_grad_fun(params, *args, **kwargs)
+      aux = None
```

Comment:
Indirectly related: a cleaner way to handle this would be to setup `self.fun_with_aux` based on `has_aux`. Then you would only need to branch based on aux once, in the setup method.